adala/utils/llm.py

```diff
@@ -0,0 +1,235 @@
+    model: Optional[str] = 'gpt-4o-mini',
+    instruction_first: Optional[bool] = True,
+    response_model: Optional[Type[BaseModel]] = None,
+    api_key: Optional[str] = None,
+    max_tokens: Optional[int] = 1000,
+    temperature: Optional[float] = 0.0,
+    timeout: Optional[Union[float, int]] = None,
+) -> Dict[str, Any]:
+
+    if not user_prompt and not messages:
+        raise ValueError("You must provide either `user_prompt` or `messages`.")
+
+    if not messages:
+        # get messages from user_prompt and system_prompt
+        messages = get_messages(user_prompt, system_prompt, instruction_first)
+    response = LLMResponse()
+
+    if response_model is None:
+        # unconstrained generation - return raw completion text and store it in `data` field: {"text": completion_text}
+        # TODO: this branch can be considered as deprecated at some point, as we always want to run LLM constrained by pydantic model  # noqa
+        try:
+            completion = litellm.completion(
+                model=model,
+                api_key=api_key,
+                messages=messages,
+                max_tokens=max_tokens,
+                temperature=temperature,
+                timeout=timeout,
+            )
+            completion_text = completion.choices[0].message.content
+            response.data['_completion_text'] = completion_text
+        except Exception as e:
+            response.adala_error = True
+            response.adala_message = type(e).__name__
+            # create a traceback string
+            response.adala_details = traceback.format_exc()
+        finally:
+            return response.model_dump(by_alias=True)
+
+    # constrained generation branch - use `response_model` to constrain the LLM response
+    try:
+        instructor_response, completion = instructor_client.chat.completions.create_with_completion(
+            model=model,
+            api_key=api_key,
+            messages=messages,
+            max_tokens=max_tokens,
+            timeout=timeout,
+            temperature=temperature,
+            response_model=response_model,
+        )
+        response.data = instructor_response.model_dump(by_alias=True)
+    except Exception as e:
+        response.adala_error = True
+        response.adala_message = type(e).__name__
+        response.adala_details = traceback.format_exc()
+    finally:
+        return response.model_dump(by_alias=True)
+
+
+def parallel_get_llm_response(
+    user_prompts: List[str],
+    system_prompt: Optional[str] = None,
+    model: Optional[str] = 'gpt-4o-mini',
+    instruction_first: Optional[bool] = True,
+    response_model: Optional[Type[BaseModel]] = None,
+    api_key: Optional[str] = None,
+    max_tokens: Optional[int] = 1000,
+    temperature: Optional[float] = 0.0
+):
+    pool = mp.Pool(mp.cpu_count())
+    responses = pool.starmap(
+        get_llm_response,
+        [
+            (
+                user_prompt,
+                system_prompt,
```

Comment:
since this is passing args positionally, wouldn't `model` be passed to `messages` in this case?

Reply:
fixed